pkg/controller/atlasdatabaseuser/atlasdatabaseuser_controller.go

```diff
@@ -78,155 +76,212 @@ func (r *AtlasDatabaseUserReconciler) Reconcile(context context.Context, req ctr
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+
 		return result.ReconcileResult(), nil
 	}
 
 	connection, err := atlas.ReadConnection(log, r.Client, r.GlobalAPISecret, project.ConnectionSecretObjectKey())
 	if err != nil {
-		result := workflow.Terminate(workflow.AtlasCredentialsNotProvided, err.Error())
-		ctx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+		result = workflow.Terminate(workflow.AtlasCredentialsNotProvided, err.Error())
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+
 		return result.ReconcileResult(), nil
 	}
-	ctx.Connection = connection
+	workflowCtx.Connection = connection
 
 	atlasClient, err := atlas.Client(r.AtlasDomain, connection, log)
 	if err != nil {
-		result := workflow.Terminate(workflow.Internal, err.Error())
-		ctx.SetConditionFromResult(status.DeploymentReadyType, result)
+		result = workflow.Terminate(workflow.Internal, err.Error())
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+
 		return result.ReconcileResult(), nil
 	}
-	ctx.Client = atlasClient
+	workflowCtx.Client = atlasClient
+
+	owner, err := customresource.IsOwner(databaseUser, r.ObjectDeletionProtection, managedByOperator(), managedByAtlas(ctx, atlasClient, project.ID(), log))
+	if err != nil {
+		result = workflow.Terminate(workflow.Internal, fmt.Sprintf("enable to resolve ownership for deletion protection: %s", err))
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+		log.Error(result.GetMessage())
 
-	result = r.ensureDatabaseUser(ctx, *project, *databaseUser)
-	if !result.IsOk() {
-		ctx.SetConditionFromResult(status.DatabaseUserReadyType, result)
 		return result.ReconcileResult(), nil
 	}
 
-	ctx.SetConditionTrue(status.DatabaseUserReadyType)
-	ctx.SetConditionTrue(status.ReadyType)
-	return result.ReconcileResult(), nil
-}
+	if !owner {
+		result = workflow.Terminate(
+			workflow.AtlasDeletionProtection,
+			"unable to reconcile database user: it already exists in Atlas, it was not previously managed by the operator, and the deletion protection is enabled.",
+		)
+		workflowCtx.SetConditionFromResult(status.DatabaseUserReadyType, result)
+		log.Error(result.GetMessage())
 
-func (r *AtlasDatabaseUserReconciler) readProjectResource(user *mdbv1.AtlasDatabaseUser, project *mdbv1.AtlasProject) workflow.Result {
-	if err := r.Client.Get(context.Background(), user.AtlasProjectObjectKey(), project); err != nil {
-		return workflow.Terminate(workflow.Internal, err.Error())
+		return result.ReconcileResult(), nil
 	}
-	return workflow.OK()
-}
 
-func (r *AtlasDatabaseUserReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	c, err := controller.New("AtlasDatabaseUser", mgr, controller.Options{Reconciler: r})
-	if err != nil {
-		return err
+	deletionRequest, result := r.handleDeletion(ctx, databaseUser, project, atlasClient, log)
+	if deletionRequest {
+		return result.ReconcileResult(), nil
```

Comment:
I am having difficulties following the logic of this reconcile function.
Why do we handle deletion here? I do not see where we are checking that a deletion is needed or requested.

Reply:
Here is where it's validated

https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1003/files#diff-b936aa2d7b65922a3ed27d3debce3bf9cb8848177e170dbf3294d4157ee22c12R202-R204

Reply:
ok, I guess the confusion was due to the name of the method, `handleDeletion` makes you think that deletion is already decided.